File: assignment 2/Frogger/Frogger/Flare.cpp
#include "Flare.h"

#define WinX 640
#define WinY 480

#define HEIGHTFROMWIDTH(w)  ((320*(w)*WinY)/(240*WinX))

#define isqrt(x)        (int)((double)(x))

Flare::Flare(VSResSurfRevLib * mySurf, VSShaderLib * shader, MathLib* calc, int lx, int ly,
		int cx, int cy)
		: StaticObject(mySurf, shader, calc), _lx(lx), _ly(ly), _cx(cx), _cy(cy) 
{
	amb[0] = 0.0f; amb[1] = 0.0f; amb[2] = 1.f; amb[3] = 1.f;
	diff[0] = .1f; diff[1] = 0.1f; diff[2] = .9f; diff[3] = .4f;
	spec[0] = 0.04f; spec[1] = 0.04f; spec[2] = .9f; spec[3] = 1.f;
	emissive[0] = 0.f;  emissive[1] = 0.f; emissive[2] = 0.f; emissive[3] = 1.f;
	shininess = new float(12.f * 128.f);
}


Flare::~Flare()
{
}

void
Flare::randomize(int nTextures, int nPieces, float fMaxSize, unsigned int minColour, unsigned int maxColour)
{
	int             i;
	float           fFracDist;
	float           fEnvelopeSize;
	FLARE_ELEMENT_DEF *element;

	// Initialize flare with requested number of elements, with
	// random placement, colour, and size of elements.
	_nPieces = nPieces;
	_fMaxSize = fMaxSize;
	_minColour = minColour;
	_maxColour = maxColour;
	fFracDist = 1.0f / (float)(_nPieces - 1);

	for (i = 0; i < _nPieces; ++i)
	{
		element = &_element[i];
		element->fDistance = (fFracDist*i) + FLARE_FRANGE(0, fFracDist);

		// Envelope size is maximum at ends of line, minimum in the middle (i.e. two
		// cones, touching at the tips).
		fEnvelopeSize = (float)fabs(1.0f - 2 * element->fDistance);

		element->fSize = FLARE_FRANGE(0.6f, 1.0f) * fEnvelopeSize;
		element->argb = FLARE_RANGE(minColour & 0xff000000, maxColour & 0xff000000) |
			FLARE_RANGE(minColour & 0x00ff0000, maxColour & 0x00ff0000) |
			FLARE_RANGE(minColour & 0x0000ff00, maxColour & 0x0000ff00) |
			FLARE_RANGE(minColour & 0x000000ff, maxColour & 0x000000ff);
	}
}

void
Flare::create(){

	int     dx, dy;          // Screen coordinates of "destination"
	int     px, py;          // Screen coordinates of flare element
	int     maxflaredist, flaredist, flaremaxsize, flarescale;
	int     width, height, alpha;    // Piece parameters;
	int     i;
	FLARE_ELEMENT_DEF    *element;
	int idAux = id;

	// Compute how far off-center the flare source is.
	maxflaredist = isqrt(_cx*_cx + _cy*_cy);
	flaredist = isqrt((_lx - _cx)*(_lx - _cx) +
		(_ly - _cy)*(_ly - _cy));
	flaredist = abs(maxflaredist - flaredist);
	flaremaxsize = (int)(WinX * _fMaxSize);
	flarescale = (int)(WinX * _fMaxSize);

	// Destination is opposite side of centre from source
	dx = _cx + (_cx - _lx);
	dy = _cy + (_cy - _ly);

	// Render each element.
	for (i = 0; i < _nPieces; ++i)
	{
		element = &_element[i];

		// Position is interpolated along line between start and destination.
		px = (int)((1.0f - element->fDistance)*_lx + element->fDistance*dx);
		py = (int)((1.0f - element->fDistance)*_ly + element->fDistance*dy);

		// Piece size are 0 to 1; flare size is proportion of
		// screen width; scale by flaredist/maxflaredist.
		width = (int)((flaredist*flarescale*element->fSize) / maxflaredist);
		
		// Width gets clamped, to allows the off-axis flares
		// to keep a good size without letting the elements get
		// too big when centered.
		if (width > flaremaxsize)
		{
			width = flaremaxsize;
		}

		// Flare elements are square (round) so height is just
		// width scaled by aspect ratio.
		height = ((320 * (width)*WinX) / (240 * WinY));
		alpha = (flaredist*(element->argb >> 24)) / maxflaredist;

		if (width > 1)
		{
			unsigned int    argb = (alpha << 24) | (element->argb & 0x00ffffff);
			_mySurf->setObjId(idAux);
			_mySurf->createRectangle(px - width / 2, py - height / 2, width, height);
			applyColor();
			idAux++;
		}
		else{ aux++; }
	}
}

void
Flare::draw()
{
	int idAux = id;
	for (int i = 0; i < _nPieces - aux; i++)
	{
		_mySurf->setObjMaterials(idAux, _shader);
		_calc->translation(_position.getX(), _position.getY() + 10.0f, _position.getZ());
		sendToGL(idAux);
		idAux++;
	}
	
}

File: DynamicObj/Flare.js
/* 
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

var elements = [];
var _fMaxSize = 0.0;
var _nPieces = 0.0;
var WinX = window.innerWidth;
var WinY = window.innerHeight;

var cam = new THREE.OrthographicCamera(0, WinX, 0, WinY, -500, 500);
var proj = cam.projectionMatrix;

// Could be a problem
function flareRange(a, b) {
    return ((Math.random() & 0xffffff) / 0xfffffe) * (b - a) + a;
}

function elementsFlare() {
    this.fDistance = 0.0;
    this.size = 0.0;
}

function randomize(nPieces, fMaxSize) {

    var fFracDist;
    var fEnvelopeSize;
    var el;

    _fMaxSize = fMaxSize;
    _nPieces = nPieces;

    fFracDist = 1.0 / (_nPieces - 1);

    for (i = 0; i < nPieces; ++i)
    {
        el = new elementsFlare();
        el.fDistance = fFracDist * i + flareRange(0, fFracDist);

        fEnvelopeSize = Math.abs(1.0 - 2.0 * el.fDistance);
        el.size = flareRange(0.6, 1.0) * fEnvelopeSize;
        elements.push(el);
    }
}

function Flare(_lx, _ly) {

    var flareImage = THREE.ImageUtils.loadTexture('particle.png');
    flareImage.magFilter = THREE.NearestFilter;

    var uniforms = {
        ambient: {type: "c", value: new THREE.Color(0x000000)},
        diffuse: {type: "c", value: new THREE.Color(0x000000)},
        spec: {type: "c", value: new THREE.Color(0x000000)},
        shininess: {type: "f", value: 32},
        billboard: {type: "f", value: 1.0},
        spotDir: {type: "v3", value: sDir},
        spotPos: {type: "v3", value: sPos},
        night: {type: "v3", value: nightMode},
        texMode: {type: "f", value: 1.0},
        texture1: {type: "t", value: flareImage},
        lens: {type: "f", value: 1.0},
        lensProj: {type: "m4", value: proj}
    };

    var material = new THREE.ShaderMaterial({vertexShader: vertexShader, fragmentShader: fragmentShader, uniforms: uniforms});

    var dx, dy;          // Screen coordinates of "destination"
    var px, py;          // Screen coordinates of flare element
    var maxflaredist, flaredist, flaremaxsize, flarescale;
    var width, height;
    var _cx = WinX / 2;
    var _cy = WinY / 2;

    // Compute how far off-center the flare source is.
    maxflaredist = _cx * _cx + _cy * _cy;
    flaredist = (_lx - _cx) * (_lx - _cx) +
            (_ly - _cy) * (_ly - _cy);
    flaredist = maxflaredist - flaredist;
    flaremaxsize = WinX * _fMaxSize;
    flarescale = WinX * _fMaxSize;

    // Destination is opposite side of centre from source
    dx = _cx + (_cx - _lx);
    dy = _cy + (_cy - _ly);

    for (i = 0; i < _nPieces; i++) {
        var el = elements[i];

        material.depthTest = false;
        material.transparent = true;
        material.blending = THREE[ "AdditiveBlending" ];

        // Position is interpolated along line between start and destination.
        px = 1.0 - el.fDistance * _lx + el.fDistance * dx;
        py = 1.0 - el.fDistance * _ly + el.fDistance * dy;

        // Piece size are 0 to 1; flare size is proportion of
        // screen width; scale by flaredist/maxflaredist.
        width = (flaredist * flarescale * el.size) / maxflaredist;

        // Width gets clamped, to allows the off-axis flares
        // to keep a good size without letting the elements get
        // too big when centered.
        if (width > flaremaxsize)
        {
            width = flaremaxsize;
        }

        var ratio = WinX / WinY;

        height = width / ratio;

        if (width > 1)
        {
            var geometry = new THREE.BoxGeometry(height, height, 0.0);
            this.flare = new THREE.Mesh(geometry, material);
            this.flare.frustumCulled = false;
            this.flare.position.x = px;
            this.flare.position.y = py;
            this.flare.position.z = 200.0;
            this.flare.renderDepth = 0.0;
            scene.add(this.flare);
        }
        
        material.depthTest = true;

    }
}





